### PR TITLE
feat: directional move focus #55

### PR DESCRIPTION
**Pull request by denismaxim0v**

### CODE REVIEW

**Comment by denismaxim0v** [comment]
that's still wip, I'll write the tests later on today

**Comment by henil** [comment]
I have tried the branch locally and it's working great for me. Just leaving some nitpicks which came to my mind from the first look.

File: src/screen.rs [hunk 1]
```diff
@@ -1282,6 +1287,154 @@ impl Screen {
...
+                        self.active_terminal = Some(active.pid);
+                        ()
+                    }
+                }
+            }
+            None => {
+                self.active_terminal = Some(active_terminal.unwrap().pid);
+                ()
+            }
+        }
```
Comment (henil) [inline src/screen.rs:1398]: .

File: src/screen.rs [hunk 2]
```diff
@@ -1282,6 +1287,154 @@ impl Screen {
...
```
Comment (henil) [inline src/screen.rs:1435]: .

File: src/screen.rs [hunk 3]
```diff
@@ -1282,6 +1287,154 @@ impl Screen {
...
```
Comment (henil) [inline src/screen.rs:1361]: Same here

File: src/screen.rs [hunk 4]
```diff
@@ -1282,6 +1287,154 @@ impl Screen {
...
```
Comment (henil) [inline src/screen.rs:1324]: I think this `match` statements should be reduced to `if let`, as it improves readability and that's what `if let` statements are for. And also reduces a level of indent :)

**Comment by imsnif** [comment]
Hey @denismaxim0v - on the surface this looks great. Before I go deeper into it, do you think you can add some tests for these? There are quite some edge cases here that we should probably be testing. (eg. making sure it picks the terminal with the largest overlap in each case, etc.) You can look at some of the resize tests in this repo for inspiration.

**Comment by denismaxim0v** [comment]
> Hey @denismaxim0v - on the surface this looks great. Before I go deeper into it, do you think you can add some tests for these? There are quite some edge cases here that we should probably be testing. (eg. making sure it picks the terminal with the largest overlap in each case, etc.) You can look at some of the resize tests in this repo for inspiration.

Yup, working on it, gonna commit those tomorrow

File: src/screen.rs (hunk 1 above)
Comment (denismaxim0v) [reply src/screen.rs:1398]: Agree, this can be reduced

File: src/screen.rs [hunk 5]
```diff
@@ -1282,6 +1287,130 @@ impl Screen {
...
+                .max_by_key(|(_, (_, c))| c.get_horizontal_overlap_with(&active))
+                .map(|(_, (pid, _))| pid);
+            match next_index {
+                Some(p) => {
+                    self.active_terminal = Some(*p);
+                }
+                None => {
+                    self.active_terminal = Some(active.pid);
+                }
+            }
```
Comment (henil) [inline src/screen.rs:1315]: Thanks for fixing the previous ones, perhaps these inner ones can too be reduced.

File: src/screen.rs [hunk 6]
```diff
@@ -1282,6 +1287,130 @@ impl Screen {
...
+                .max_by_key(|(_, (_, c))| c.get_vertical_overlap_with(&active))
+                .map(|(_, (pid, _))| pid);
+            match next_index {
+                Some(p) => {
+                    self.active_terminal = Some(*p);
+                }
+                None => {
+                    self.active_terminal = Some(active.pid);
+                }
+            }
```
Comment (henil) [inline src/screen.rs:1346]: .

File: src/screen.rs [hunk 7]
```diff
@@ -1282,6 +1287,130 @@ impl Screen {
...
```
Comment (henil) [inline src/screen.rs:1377]: .

File: src/screen.rs [hunk 8]
```diff
@@ -1282,6 +1287,130 @@ impl Screen {
...
+                .max_by_key(|(_, (_, c))| c.get_horizontal_overlap_with(&active))
+                .map(|(_, (pid, _))| pid);
+            match next_index {
+                Some(p) => {
+                    self.active_terminal = Some(*p);
+                }
+                None => {
+                    self.active_terminal = Some(active.pid);
+                }
+            }
```
Comment (henil) [inline src/screen.rs:1408]: .

**Review by henil — CHANGES_REQUESTED** [review]

File: src/terminal_pane/terminal_pane.rs [hunk 9]
```diff
@@ -164,6 +164,48 @@ impl TerminalPane {
...
+                && (self.get_x() + self.get_columns() >= (other.get_x() + other.get_columns())))
+            || (other.get_x() <= self.get_x()
+                && (other.get_x() + other.get_columns() >= (self.get_x() + self.get_columns())))
+    }
+    pub fn get_vertical_overlap_with(&self, other: &TerminalPane) -> usize {
+        std::cmp::min(
+            self.get_x() + self.get_columns(),
+            other.get_x() + other.get_columns(),
+        ) - std::cmp::max(self.get_x(), other.get_x())
+    }
```
Comment (imsnif) [inline src/terminal_pane/terminal_pane.rs:208]: Do you think we can implement these on the `Rect` trait instead of on terminal_pane? That way when we add other pane types (eg. plugins) they will get this for free.

File: src/tests/integration/move_focus_down.rs [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    let mut fake_input_output = get_fake_os_input(&fake_win_size);
+    fake_input_output.add_terminal_input(&[
+        &COMMAND_TOGGLE,
+        &COMMAND_TOGGLE,
+        &SPLIT_HORIZONTALLY,
+        &SPLIT_VERTICALLY,
+        &SPLIT_VERTICALLY,
+        &MOVE_FOCUS_UP,
+        &MOVE_FOCUS_DOWN,
+        &QUIT,
```
Comment (imsnif) [inline src/tests/integration/move_focus_down.rs:64]: Correct me if I'm wrong, but I think here both pane candidates are of the same size, aren't they? Maybe we can adjust the size of one of them using a resize command in this input? (this is true to the other tests as well, I think)

**Review by imsnif — CHANGES_REQUESTED** [review]
Very cool! I'm looking forward to this feature :)

Great work. The code looks very clean and the tests are great. I left two relatively minor comments and after addressing them I think we can merge this.

File: src/tests/integration/move_focus_down.rs (hunk 10 above)
Comment (denismaxim0v) [reply src/tests/integration/move_focus_down.rs:64]: Yup, my bad, will push it with a resize

File: src/terminal_pane/terminal_pane.rs (hunk 9 above)
Comment (denismaxim0v) [reply src/terminal_pane/terminal_pane.rs:208]: Yup, don't see why it can't be moved, will do

File: src/tests/integration/move_focus_down.rs (hunk 10 above)
Comment (denismaxim0v) [reply src/tests/integration/move_focus_down.rs:64]: Now that I look at it I do the vertical split two times so the panes are 25%, 25%, 50% so it's still testing whether it switches to the most overlapping one

Comment (imsnif) [reply src/tests/integration/move_focus_down.rs:64]: You're right. My bad :)

**Review by imsnif — APPROVED** [review]
Nice solution with `Self` - I did not know you could do that :)
